### PR TITLE
MachineImage per worker

### PR DESCRIPTION
**Pull request by KristianZH**
**What this PR does / why we need it**:
The MachineImage will be taken from the worker pool. With this change every workerPool can have different MachineImage.

If MachineImage is not set on worker level, it will be taken from the cloud(as before).

**Which issue(s) this PR fixes**:
Fixes #

```noteworthy user
Prior releases of Gardener only allowed exactly one machine image/version per `Shoot`. It is now possible to configure different machine images *per worker pool*. The `.spec.cloud.<provider-name>.machineImage` section contains the default image that should be used for worker pools which don't explicitly configure a machine image (which is possible via the `.spec.cloud.<provider-name>.workers[*].machineImage` field).
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Let's first merge #1177 and then you can rebase on top of it.

**Comment by rfranzke** [comment]
@KristianZH #1177 is merged, please rebase now.

File: test/integration/shoots/worker/worker_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package worker
```
Comment (rfranzke) [inline test/integration/shoots/worker/worker_suite_test.go:15]: license header is missing

File: pkg/apis/garden/v1beta1/helper/helpers.go [hunk 2]
```diff
@@ -159,8 +160,8 @@ func GetShootCloudProviderWorkers(cloudProvider gardenv1beta1.CloudProvider, sho
 	return workers
 }
 
-// GetMachineImageFromShoot returns the machine image used in a shoot manifest, however, it requires the cloud provider as input.
-func GetMachineImageFromShoot(cloudProvider gardenv1beta1.CloudProvider, shoot *gardenv1beta1.Shoot) *gardenv1beta1.ShootMachineImage {
+// GetCloudMachineImageFromShoot returns the machine image used in a shoot manifest, however, it requires the cloud provider as input.
+func GetCloudMachineImageFromShoot(cloudProvider gardenv1beta1.CloudProvider, shoot *gardenv1beta1.Shoot) *gardenv1beta1.ShootMachineImage {
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/helper/helpers.go]: Why is this renamed?

File: pkg/apis/garden/v1beta1/helper/helpers.go [hunk 3]
```diff
@@ -255,7 +290,7 @@ func SetMachineImages(profile *gardenv1beta1.CloudProfile, images []gardenv1beta
 	return nil
 }
 
-// SetMachineImages sets imageVersions to the matching imageName in the machineImages.
+// SetMachineImageVersionsToMachineImage sets imageVersions to the matching imageName in the machineImages.
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/helper/helpers.go:293]: Why is this renamed as well?

File: pkg/operation/shoot/shoot.go [hunk 4]
```diff
@@ -266,9 +266,14 @@ func (s *Shoot) GetNodeNetwork() gardencorev1alpha1.CIDR {
 	return *k8sNetworks.Nodes
 }
 
-// GetMachineImage returns the name of the used machine image.
-func (s *Shoot) GetMachineImage() *gardenv1beta1.ShootMachineImage {
-	return helper.GetMachineImageFromShoot(s.CloudProvider, s.Info)
+// GetCloudMachineImage returns the name of the used machine image.
+func (s *Shoot) GetCloudMachineImage() *gardenv1beta1.ShootMachineImage {
```
Comment (rfranzke) [inline pkg/operation/shoot/shoot.go]:
```suggestion
func (s *Shoot) GetDefaultMachineImage() *gardenv1beta1.ShootMachineImage {
```

What I've understood is that if the `.spec.cloud.<provider>.machineImage` field is set then this value is used as default for all worker pools that don't explicitly configure a machine image themselves.

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 5]
```diff
@@ -221,45 +221,40 @@ func hasMaintainNowAnnotation(shoot *gardenv1beta1.Shoot) bool {
...
-	if !versionExistsInCloudProfile || *shoot.Spec.Maintenance.AutoUpdate.MachineImageVersion || ForceUpdateRequired(shootCurrentImage, machineImageFromCloudProfile) {
-		return updateToLatestMachineImageVersion(machineImageFromCloudProfile)
-	}
+	for _, shootImage := range shootCurrentImages {
+		machineImagesFound, machineImageFromCloudProfile, err := helper.DetermineMachineImageForName(*cloudProfile, shootImage.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failure while determining the machine image in the CloudProfile: %s", err.Error())
+		}
+		if !machineImagesFound {
+			return nil, fmt.Errorf("failure while determining the machine image in the CloudProfile: no machineImage with name '%s' (specified in shoot) could be found in the cloud cloudProfile '%s'", shootImage.Name, cloudProfile.Name)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]:
```suggestion
			return nil, fmt.Errorf("failure while determining the machine image in the CloudProfile: no machineImage with name '%s' (specified in shoot) could be found in the cloudProfile '%s'", shootImage.Name, cloudProfile.Name)
```

File: pkg/apis/garden/v1beta1/helper/helpers.go (hunk 2 above)
Comment (KristianZH) [reply pkg/apis/garden/v1beta1/helper/helpers.go]: Because this functions returns the machine image which is in the Spec -> Cloud -> CloudName -> MachineImage, not the actual machineImages which are used by the Worker(Spec -> Cloud -> CloudName -> []Workers -> []MachineImages)

File: pkg/operation/shoot/shoot.go (hunk 4 above)
Comment (KristianZH) [reply pkg/operation/shoot/shoot.go]: Yes, correct. If .spec.cloud.<provider>.machineImage is not set, it comes with default machineImage.

File: pkg/apis/garden/v1beta1/helper/helpers.go (hunk 2 above)
Comment (rfranzke) [reply pkg/apis/garden/v1beta1/helper/helpers.go]: OK, let's call it `GetDefaultMachineImageFromShoot` then. The `worker` section is also below `.spec.cloud.<...>`.

File: pkg/apis/garden/v1beta1/helper/helpers.go (hunk 3 above)
Comment (KristianZH) [reply pkg/apis/garden/v1beta1/helper/helpers.go:293]: Because I think this name better fits on the functions's behaviour. I can revert it, if you think the old one is better :)

Comment (rfranzke) [reply pkg/apis/garden/v1beta1/helper/helpers.go:293]: Just checked the function again, it's fine. Thanks! :)

File: pkg/apis/garden/v1beta1/helper/helpers.go (hunk 2 above)
Comment (KristianZH) [reply pkg/apis/garden/v1beta1/helper/helpers.go]: Okay

**Comment by rfranzke** [comment]
The test is failing now, PTAL.

**Comment by KristianZH** [comment]
Fixed

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 6]
```diff
@@ -158,15 +158,15 @@ func (c *defaultMaintenanceControl) Maintain(shootObj *gardenv1beta1.Shoot, key
...
 		// continue execution to allow the kubernetes version update
 		handleError(fmt.Sprintf("Could not maintain machine image version: %s", err.Error()))
 	}
 
-	var updateMachineImage func(s *gardenv1beta1.Cloud)
-	if updateImage {
-		updateMachineImage = helper.UpdateMachineImage(operation.Shoot.CloudProvider, machineImage)
+	var updateMachineImages func(s *gardenv1beta1.Cloud)
+	if len(machineImages) > 0 {
+		updateMachineImages = helper.UpdateMachineImages(operation.Shoot.CloudProvider, machineImages)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]: This only updates the machine images in the worker section. The default machine image in `.spec.cloud.<provider>.machineImage` is not maintained anymore. This needs to be changed back.

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Review by rfranzke — APPROVED** [review]
/lgtm
